controller/identity/creds_watcher.go

```diff
@@ -0,0 +1,97 @@
+package identity
+
+import (
+	"crypto/x509"
+	"fmt"
+	"path/filepath"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/linkerd/linkerd2/pkg/tls"
+	log "github.com/sirupsen/logrus"
+)
+
+const dataDirectoryLnName = "..data"
+
+// FsCredsWatcher is used to monitor tls credentials on the filesystem
+type FsCredsWatcher struct {
+	issuerPath, keyName, crtName, expectedName string
+	roots                                      *x509.CertPool
+	validity                                   tls.Validity
+	issuerChan                                 chan tls.Issuer
+}
+
+// NewFsCredsWatcher constructs a FsCredsWatcher instance
+func NewFsCredsWatcher(issuerPath, keyName, crtName, expectedName string, roots *x509.CertPool, validity tls.Validity) *FsCredsWatcher {
+	ch := make(chan tls.Issuer, 100)
+	return &FsCredsWatcher{issuerPath, keyName, crtName, expectedName, roots, validity, ch}
+}
+
+// Creds gives back a chan from which new issuers can be read
+func (fscw *FsCredsWatcher) Creds() <-chan tls.Issuer {
+	return fscw.issuerChan
+}
+
+func (fscw *FsCredsWatcher) loadCredentials() (*tls.CA, error) {
+	creds, err := tls.ReadPEMCreds(
+		filepath.Join(fscw.issuerPath, fscw.keyName),
+		filepath.Join(fscw.issuerPath, fscw.crtName),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read CA from %s: %s", fscw.issuerPath, err)
+	}
+
+	if err := creds.Crt.Verify(fscw.roots, fscw.expectedName); err != nil {
+		return nil, fmt.Errorf("failed to verify issuer credentials for '%s' with trust anchors: %s", fscw.expectedName, err)
+	}
+
+	log.Infof("Loaded issuer cert:\nCert: %s\nKey: %s", creds.EncodeCertificatePEM(), creds.EncodePrivateKeyPEM())
+	return tls.NewCA(*creds, fscw.validity), nil
+}
+
+// StartWatching starts watching the filesystem for cert updates
+func (fscw *FsCredsWatcher) StartWatching() error {
```

Comment:
I wonder if we are trying to do too much in this method. How about instead of sending `*tls.CA` to the channel, we just send it an "event", and then let `identity.Service` do most of the TLS-related stuff? 

I think there might be a way to make this and the changes to `pkg/identity/services.go` simpler. I am experimenting with it in this [gist](https://gist.github.com/ihcsim/8580c88f94fffc0c345fc852e4bb38ae). The initializing and switching of the `issuer` object will be interesting.

Reply:
Looked at the diff. I though about something similar in the beginning. We can go both ways but my initial idea was that this makes the service harder to test in case we want to add some extensive testing to it. 

Since the service is has quite a bit of logic in there that has to do with issuing certs, we might want to exercise some of these scenarios by having it call into `tls.CA` created from different certs. If that is the case, we will have to put the issuer certs on disk and have different ones for every test case while pinging the service some dummy event. Supplying the `tls.CA` via the channel allows us to mock it in a much easier way I think. 

Wrt to initializing and swapping the `tls.CA`, I dont see much of a problem as long as the lock is held. Maybe I am missing smth here? In any way, we can go either direction. Do not really have a strong opinion.

Reply:
Whats more interesting to me however is all the error handling in case we have errors from the error channel or invalid issuer certs written to the file system

Reply:
The benefit of moving the TLS code out of the CredsWatcher is that we can convert it into a generic file watcher package, say `pkg/watcher/file`.

In terms of testing the `identity.Service`, I think certain unexported functions (like `loadCredentials` if we choose to move it out of the file watcher) can be mock by using function pointers. There are some examples in `cli/cmd/install.go`. Testing `Certify()` will be difficult even in its current state.

> If that is the case, we will have to put the issuer certs on disk and have different ones for every test case while pinging the service some dummy event. 

Sounds like a good integration test. We can create a separate issue for that, as I assume there will be quite a bit of new test code.

> Whats more interesting to me however is all the error handling in case we have errors from the error channel or invalid issuer certs written to the file system

Good point. My first thought is first of all, the file watcher shouldn't worry about how to handle it; the identity service should. Then, the old cert shouldn't be rotated and the identity service shouldn't go down. As to how we notify the users in addition to generate logs, we can consider publishing k8s events. I don't know if we need to tackle this here, but controller/proxy-injector/webhook.go has some good examples on recording k8s events.

Reply:
These are all great points. Thanks!